File: src/core/collectors/claude-code.js
'use strict';

const { t } = require('../../i18n');

const path = require('path');
const fs = require('fs');
const { walkFiles, readJsonlFrom, projectName, homeDir } = require('../util');

/**
 * Collecteur Claude Code : lit les fichiers .jsonl sous ~/.claude/projects.
 *
 * Chaque message d'assistant y porte un bloc `usage` complet, y compris la
 * ventilation du cache par TTL (`ephemeral_5m` / `ephemeral_1h`), ce qui
 * permet une tarification exacte plutôt qu'approchée.
 *
 * Deux pièges dans ce format, tous deux traités ici :
 *
 *  1. Un même message apparaît plusieurs fois — Claude Code réécrit la ligne
 *     au fil du streaming. Les doublons partagent `message.id`, et les
 *     compter deux fois doublerait purement et simplement la facture affichée.
 *  2. Les entrées `quotaLimits` (émises sur 429) portent l'état réel des
 *     limites de débit : type de fenêtre et instant de réinitialisation. C'est
 *     la seule source fiable sur le sujet, on la récupère au passage.
 */

const SOURCE = 'claude-code';

function rootDir(config = {}) {
  return config.claudeCodeDir || path.join(homeDir(), '.claude', 'projects');
}

function isAvailable(config = {}) {
  try {
    return fs.statSync(rootDir(config)).isDirectory();
  } catch {
    return false;
  }
}

/**
 * Classe la CAUSE réelle d'un refus.
 *
 * `rateLimitType` vaut toujours « five_hour » : il désigne la fenêtre dont on
 * rapporte l'heure de réinitialisation, PAS ce qui a bloqué la requête. Sur ce
 * poste, deux refus sur trois viennent en réalité d'un plafond de dépense
 * mensuel. Les confondre revient à calibrer la jauge 5 h sur un événement qui
 * n'a rien à voir avec elle.
 */
function rejectionCause(rec) {
  const text = (((rec.message || {}).content) || [])
    .map((c) => (typeof c === 'string' ? c : c.text || ''))
    .join(' ')
    .toLowerCase();

  if (/spend limit|spending limit|credit balance|out of credits/.test(text)) return 'spend';
  if (/weekly limit/.test(text)) return 'weekly';
  if (/session limit|usage limit|rate limit/.test(text)) return 'window';
  return 'unknown';
}

/** Convertit un bloc `usage` brut en structure de tokens canonique. */
function extractTokens(usage) {
  const cc = usage.cache_creation || {};
  const w5 = cc.ephemeral_5m_input_tokens || 0;
  const w1 = cc.ephemeral_1h_input_tokens || 0;
  const input = usage.input_tokens || 0;
  const output = usage.output_tokens || 0;
  const cacheRead = usage.cache_read_input_tokens || 0;
  // On préfère la ventilation par TTL quand elle est là ; `cache_creation_input_tokens`
  // sert de repli pour les versions de logs qui ne la fournissent pas.
  const cacheWrite = w5 + w1 || usage.cache_creation_input_tokens || 0;
  const thinking = (usage.output_tokens_details && usage.output_tokens_details.thinking_tokens) || 0;

  return {
    input,
    output,
    cacheRead,
    cacheWrite,
    cacheWrite5m: w5 || (w1 ? 0 : cacheWrite),
    cacheWrite1h: w1,
    thinking,
    total: input + output + cacheRead + cacheWrite,
  };
}

/**
 * Scanne les fichiers de logs, en repartant de l'état d'indexation fourni.
 *
 * @param {object} state  index persistant : { files: { [path]: {size, offset, seen[]} } }
 * @returns {{events:Array, quota:Array, state:object, stats:object}}
 */
function collect(config = {}, state = {}) {
  const dir = rootDir(config);
  const files = walkFiles(dir, (f) => f.endsWith('.jsonl'));
  const nextState = { files: {} };
  const events = [];
  const quota = [];
  let skippedDuplicates = 0;

  for (const file of files) {
    const prev = (state.files && state.files[file]) || { offset: 0, seen: [] };
    // Fenêtre de déduplication : les réécritures de streaming sont adjacentes
    // dans le fichier, une fenêtre glissante bornée suffit et évite de garder
    // en mémoire l'intégralité des identifiants déjà vus.
    const seen = new Set(prev.seen || []);
    const order = [...(prev.seen || [])];

    const res = readJsonlFrom(file, prev.offset || 0, (rec) => {
      // --- état des limites de débit -------------------------------------
      if (rec.quotaLimits && rec.quotaLimits.resetsAt) {
        quota.push({
          source: SOURCE,
          ts: rec.timestamp ? Date.parse(rec.timestamp) : Date.now(),
          type: rec.quotaLimits.rateLimitType || 'unknown',
          status: rec.quotaLimits.status || null,
          resetsAt: rec.quotaLimits.resetsAt * 1000,
          usingOverage: !!rec.quotaLimits.isUsingOverage,
          cause: rejectionCause(rec),
        });
      }

      // --- consommation ---------------------------------------------------
      const msg = rec.message;
      if (!msg || msg.role !== 'assistant' || !msg.usage) return;

      const dedupKey = msg.id || rec.requestId;
      if (dedupKey) {
        if (seen.has(dedupKey)) {
          skippedDuplicates++;
          return;
        }
        seen.add(dedupKey);
        order.push(dedupKey);
        if (order.length > 500) seen.delete(order.shift());
      }

      const tokens = extractTokens(msg.usage);
      if (!tokens.total) return;

      events.push({
        ts: rec.timestamp ? Date.parse(rec.timestamp) : Date.now(),
        source: SOURCE,
        model: msg.model || 'unknown',
        project: projectName(rec.cwd),
        session: rec.sessionId || null,
        tokens,
        requests: 1,
      });
    });

    nextState.files[file] = { offset: res.offset, seen: order.slice(-500) };
  }

  return {
    events,
    quota,
    state: nextState,
    stats: { files: files.length, events: events.length, skippedDuplicates },
  };
}

module.exports = { id: SOURCE, get label() { return t('source.claude-code'); }, isAvailable, collect, extractTokens };

File: crates/trace-core/src/collectors/claude_code.rs
//! Collecteur Claude Code : les fichiers `.jsonl` sous `~/.claude/projects`.
//!
//! Chaque message d'assistant y porte un bloc `usage` complet, ventilation du
//! cache par TTL comprise (`ephemeral_5m` / `ephemeral_1h`), ce qui permet une
//! tarification exacte plutôt qu'approchée.
//!
//! Deux pièges dans ce format, tous deux traités ici :
//!
//!  1. Un même message apparaît plusieurs fois — Claude Code réécrit la ligne
//!     au fil du streaming. Les doublons partagent `message.id`, et les
//!     compter deux fois doublerait purement et simplement la facture.
//!  2. Les entrées `quotaLimits`, émises sur 429, portent l'état réel des
//!     limites : type de fenêtre et instant de réinitialisation. C'est la
//!     seule source fiable sur le sujet, on la récupère au passage.

use super::{
    parse_ts, Cause, Collected, CollectorState, Event, FileCursor, Quota, Stats, DEDUP_WINDOW,
};
use crate::util::{now_ms, project_name, read_jsonl_from, walk_files, Tokens};
use serde_json::Value;
use std::collections::{HashSet, VecDeque};
use std::path::{Path, PathBuf};

pub const SOURCE: &str = "claude-code";

pub fn root_dir(configured: Option<&str>) -> PathBuf {
    match configured {
        Some(d) if !d.is_empty() => PathBuf::from(d),
        _ => crate::util::home_dir().join(".claude").join("projects"),
    }
}

pub fn is_available(configured: Option<&str>) -> bool {
    root_dir(configured).is_dir()
}

/// Classe la CAUSE réelle d'un refus, à partir du texte du message.
pub fn rejection_cause(rec: &Value) -> Cause {
    let text = rec["message"]["content"]
        .as_array()
        .map(|blocks| {
            blocks
                .iter()
                .map(|c| match c {
                    Value::String(s) => s.as_str(),
                    other => other["text"].as_str().unwrap_or(""),
                })
                .collect::<Vec<_>>()
                .join(" ")
        })
        .unwrap_or_default()
        .to_lowercase();

    if ["spend limit", "spending limit", "credit balance", "out of credits"]
        .iter()
        .any(|p| text.contains(p))
    {
        return Cause::Spend;
    }
    if text.contains("weekly limit") {
        return Cause::Weekly;
    }
    if ["session limit", "usage limit", "rate limit"].iter().any(|p| text.contains(p)) {
        return Cause::Window;
    }
    Cause::Unknown
}

/// Convertit un bloc `usage` brut en structure de tokens canonique.
pub fn extract_tokens(usage: &Value) -> Tokens {
    let n = |v: &Value| v.as_i64().unwrap_or(0);

    let cc = &usage["cache_creation"];
    let w5 = n(&cc["ephemeral_5m_input_tokens"]);
    let w1 = n(&cc["ephemeral_1h_input_tokens"]);
    let input = n(&usage["input_tokens"]);
    let output = n(&usage["output_tokens"]);
    let cache_read = n(&usage["cache_read_input_tokens"]);

    // On préfère la ventilation par TTL quand elle est là ;
    // `cache_creation_input_tokens` sert de repli pour les versions de
    // journaux qui ne la fournissent pas.
    let declared = w5 + w1;
    let raw_total = if declared != 0 { declared } else { n(&usage["cache_creation_input_tokens"]) };
    let (cache_write, cache_write5m, cache_write1h) = Tokens::split_cache_write(raw_total, w5, w1);

    let thinking = n(&usage["output_tokens_details"]["thinking_tokens"]);

    Tokens {
        input,
        output,
        cache_read,
        cache_write,
        cache_write5m,
        cache_write1h,
        thinking,
        // `total` est ce que le fournisseur a facturé, pas une somme des
        // parties : les TTL sont déjà comptés dans `cache_write`.
        total: input + output + cache_read + cache_write,
    }
}

/// Scanne les journaux, en repartant de l'état d'indexation fourni.
pub fn collect(configured_dir: Option<&str>, state: &CollectorState) -> Collected {
    let dir = root_dir(configured_dir);
    let files = walk_files(&dir, &|p: &Path| {
        p.extension().and_then(|e| e.to_str()) == Some("jsonl")
    });

    let mut out = Collected {
        state: CollectorState { files: Default::default() },
        stats: Stats { files: files.len(), ..Default::default() },
        ..Default::default()
    };

    for file in &files {
        let key = file.to_string_lossy().to_string();
        let prev = state.files.get(&key).cloned().unwrap_or_default();

        let mut seen: HashSet<String> = prev.seen.iter().cloned().collect();
        let mut order: VecDeque<String> = prev.seen.iter().cloned().collect();

        let mut events = Vec::new();
        let mut quota = Vec::new();
        let mut skipped = 0usize;

        let res = read_jsonl_from(file, prev.offset, |rec| {
            // --- état des limites de débit ---------------------------------
            let limits = &rec["quotaLimits"];
            if let Some(resets_at) = limits["resetsAt"].as_f64() {
                quota.push(Quota {
                    source: SOURCE.to_string(),
                    ts: parse_ts(rec["timestamp"].as_str()).unwrap_or_else(now_ms),
                    kind: limits["rateLimitType"].as_str().unwrap_or("unknown").to_string(),
                    status: limits["status"].as_str().map(str::to_string),
                    resets_at: (resets_at * 1000.0) as i64,
                    using_overage: limits["isUsingOverage"].as_bool().unwrap_or(false),
                    cause: rejection_cause(rec),
                });
            }

            // --- consommation ----------------------------------------------
            let msg = &rec["message"];
            if msg["role"].as_str() != Some("assistant") || !msg["usage"].is_object() {
                return;
            }

            let dedup_key = msg["id"].as_str().or_else(|| rec["requestId"].as_str());
            if let Some(k) = dedup_key {
                if seen.contains(k) {
                    skipped += 1;
                    return;
                }
                seen.insert(k.to_string());
                order.push_back(k.to_string());
                if order.len() > DEDUP_WINDOW {
                    if let Some(old) = order.pop_front() {
                        seen.remove(&old);
                    }
                }
            }

            let tokens = extract_tokens(&msg["usage"]);
            if tokens.total == 0 {
                return;
            }

            events.push(Event {
                ts: parse_ts(rec["timestamp"].as_str()).unwrap_or_else(now_ms),
                source: SOURCE.to_string(),
                model: msg["model"].as_str().unwrap_or("unknown").to_string(),
                project: rec["cwd"].as_str().and_then(project_name),
                session: rec["sessionId"].as_str().map(str::to_string),
                tokens,
                requests: 1,
                compacted: None,
            });
        });

        let keep = order.len().saturating_sub(DEDUP_WINDOW);
        out.state.files.insert(
            key,
            FileCursor { offset: res.offset, seen: order.into_iter().skip(keep).collect() },
        );
        out.stats.skipped_duplicates += skipped;
        out.events.append(&mut events);
        out.quota.append(&mut quota);
    }

    out.stats.events = out.events.len();
    out
}
